### PR TITLE
Jump between sources

### PR DESCRIPTION
**Pull request by wh11e7rue**
I have implemented a feature like [unite_rotate_next_source](https://github.com/Shougo/unite.vim/blob/master/doc/unite.txt#L967)
I'm using it with init.vim like this
```vim
call denite#custom#map('insert', '<LEFT>', 'jump_to_prev_source')
call denite#custom#map('insert', '<RIGHT>', 'jump_to_next_source')
```

### CODE REVIEW

**Comment by Shougo** [comment]
OK.  I will check the implementation

File: doc/denite.txt [hunk 1]
```diff
@@ -252,6 +252,14 @@ move_to_next_line
 move_to_prev_line
 		Move to previous line.
 
+						*denite-map-jump_to_next_source*
+jump_to_next_source
```
Comment (Shougo) [inline doc/denite.txt]: Please sort the section order.

File: rplugin/python3/denite/ui/default.py [hunk 2]
```diff
@@ -491,6 +491,80 @@ def scroll_up(self, scroll):
             self.__cursor = max(self.__cursor - scroll, 0)
         else:
             return
+
+    def jump_to_next_source(self):
+        if len(self.__context['sources']) == 1:
+            return
+
+        current_line = self.__cursor + self.__win_cursor - 1
```
Comment (Shougo) [inline rplugin/python3/denite/ui/default.py]: I think it should be `current_index`.

File: rplugin/python3/denite/ui/default.py [hunk 3]
```diff
@@ -491,6 +491,80 @@ def scroll_up(self, scroll):
...
+        forward_candidates = self.__candidates[current_line:]
+        forward_sources = groupby(
+            forward_candidates,
+            lambda candidate: candidate['source']
+        )
+        forward_times = len(list(next(forward_sources)[1]))
+        remaining_candidates = self.__candidates_len - current_line \
+            - forward_times
+        if next(forward_sources, None) is None:
+            # if the cursor is on the last source
```
Comment (Shougo) [inline rplugin/python3/denite/ui/default.py]: `# If`

File: rplugin/python3/denite/ui/default.py [hunk 4]
```diff
@@ -491,6 +491,80 @@ def scroll_up(self, scroll):
...
+            self.__cursor = 0
+            self.__win_cursor = 1
+        elif self.__candidates_len < self.__winheight:
+            # if there is a space under the candidates
```
Comment (Shougo) [inline rplugin/python3/denite/ui/default.py]: `# If`

File: rplugin/python3/denite/ui/default.py [hunk 5]
```diff
@@ -491,6 +491,80 @@ def scroll_up(self, scroll):
...
+        current_line = self.__cursor + self.__win_cursor - 1
+        backward_candidates = reversed(self.__candidates[:current_line + 1])
+        backward_sources = groupby(
+            backward_candidates,
+            lambda candidate: candidate['source']
+        )
+        current_source = list(next(backward_sources)[1])
+        try:
+            prev_source = list(next(backward_sources)[1])
+        except StopIteration:  # if the cursor is on the first source
```
Comment (Shougo) [inline rplugin/python3/denite/ui/default.py]: `# If`

File: rplugin/python3/denite/ui/default.py [hunk 6]
```diff
@@ -491,6 +491,80 @@ def scroll_up(self, scroll):
...
+            self.__cursor += forward_times + self.__win_cursor - 1
+            self.__win_cursor = 1
+
+        self.update_buffer()
+
+    def jump_to_prev_source(self):
+        if len(self.__context['sources']) == 1:
+            return
+
+        current_line = self.__cursor + self.__win_cursor - 1
```
Comment (Shougo) [inline rplugin/python3/denite/ui/default.py]: current_index

File: doc/denite.txt (hunk 1 above)
Comment (wh11e7rue) [inline doc/denite.txt]: What do you mean by "section order"?

**Comment by wh11e7rue** [comment]
Thanks for reviewing.
fixed

File: doc/denite.txt (hunk 1 above)
Comment (Shougo) [inline doc/denite.txt]: The order must be alphabetical.
Please check the map sections order.

Comment (wh11e7rue) [inline doc/denite.txt]: I got it. fixed

**Comment by Shougo** [comment]
Merged.

**Pull request merged**